### DISCUSSION TITLE
Embed Gtk Window On Linux.

### DISCUSSION THREAD

**Discussion by cangunaydin** [AvaloniaUI/Avalonia#16578] [Q&A]
Hello,
I am very new on linux.
I use avalonia and created a very simple media player. Now I want to support a webview by using libwebkit2gtk on linux.
I installed ubuntu 22.04 and trying to create a very simple embedding example.
I have looked into these old issues that is related with linux.

https://github.com/AvaloniaUI/Avalonia/issues/10354
https://github.com/AvaloniaUI/Avalonia/issues/6124

and i have the same problem as the others. 
as @kekekeks  mentioned in one of the issues
i am trying to use XReparentWindow to fix the problem. but when i do that i am having
`(UrlTest:49845): Gdk-WARNING **: 17:24:38.904: The program 'UrlTest' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadWindow (invalid Window parameter)'.
  (Details: serial 313 error_code 3 request_code 7 (core protocol) minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)` 

here is a screenshot for my problem when i do not use XReparentWindow
![Screenshot from 2024-08-02 17-31-08](https://github.com/user-attachments/assets/64d0365e-5ff5-4ad2-8283-75dc3f728a2d)

and here is the code that i use.

``` csharp
using System;
using System.Runtime.InteropServices;
using System.Threading;


public class GtkWebView
{
    [DllImport("libgtk-3.so.0")]
    public static extern void gtk_init(int argc, IntPtr argv);

    [DllImport("libgtk-3.so.0")]
    public static extern IntPtr gtk_window_new(int type);

    [DllImport("libgtk-3.so.0")]
    public static extern void gtk_widget_show_all(IntPtr widget);

    [DllImport("libgtk-3.so.0")]
    public static extern void gtk_container_add(IntPtr container, IntPtr widget);

    [DllImport("libgtk-3.so.0")]
    public static extern IntPtr gtk_widget_get_display(IntPtr widget);

    [DllImport("libgtk-3.so.0")]
    public static extern void gtk_widget_realize(IntPtr widget);

    [DllImport("libgtk-3.so.0")]
    public static extern void gtk_widget_grab_focus(IntPtr widget);

    [DllImport("libgtk-3.so.0")]
    public static extern IntPtr gtk_widget_get_window(IntPtr gtkWidget);

    [DllImport("libgdk-3.so.0")]
    public static extern IntPtr gdk_x11_display_get_xdisplay(IntPtr display);

    [DllImport("libgdk-3.so.0")]
    public static extern IntPtr gdk_x11_window_get_xid(IntPtr window);

    [DllImport("libgdk-3.so.0")]
    public static extern void gtk_widget_destroy(IntPtr gtkWidget);

    [DllImport("libX11.so.6")]
    public static extern int XSync(IntPtr display, bool discard);

    [DllImport("libX11.so.6")]
    public static extern int XReparentWindow(IntPtr display, IntPtr window, IntPtr parent, int x, int y);

    [DllImport("libwebkit2gtk-4.0.so.37")]
    public static extern IntPtr webkit_web_view_new();

    [DllImport("libwebkit2gtk-4.0.so.37")]
    public static extern void webkit_web_view_load_uri(IntPtr webView, string uri);

    [DllImport("libwebkit2gtk-4.0.so.37")]
    public static extern void webkit_web_view_reload(IntPtr webView);

    private IntPtr _win;
    private IntPtr _view;

    public IntPtr Window => _win;
    public IntPtr WebView => _view;


    public GtkWebView()
    {
        gtk_init(0, IntPtr.Zero);
        //InitializeWebView();
    }

    public nint InitializeWebView(IntPtr parentWindow)
    {
        var xid = Avalonia.X11.Interop.GtkInteropHelper.RunOnGlibThread(() =>
        {
            _win = gtk_window_new(0); // GTK.WindowType.Toplevel
            _view = webkit_web_view_new();

            gtk_container_add(_win, _view);

            var display = gdk_x11_display_get_xdisplay(gtk_widget_get_display(_win));
            //XSync(display, false);

            // Thread.Sleep(10);
            gtk_widget_realize(_win);

            var xwin = gdk_x11_window_get_xid(gtk_widget_get_window(_win));
            // _ = XReparentWindow(
            //     display,
            //     xwin,
            //     parentWindow,
            //     0,
            //     0
            // );

            // XSync(display, false);

            // Thread.Sleep(10);
            gtk_widget_show_all(_win);
            gtk_widget_grab_focus(_view);
            return xwin;
        }).Result;
        return xid;
    }

    public void Navigate(string url)
    {
        Avalonia.X11.Interop.GtkInteropHelper.RunOnGlibThread(() =>
        {
            webkit_web_view_load_uri(_view, url);
            return 0;
        });
    }

    public void Reload()
    {
        Avalonia.X11.Interop.GtkInteropHelper.RunOnGlibThread(() =>
        {
            webkit_web_view_reload(_view);
            return 0;
        });
    }
}
```
and even if i do not use XReparentWindow i am getting error on the console which is.
`[IME] Unable to create IME input context:
Tmds.DBus.Protocol.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path “/org/freedesktop/IBus”
   at Tmds.DBus.Protocol.DBusConnection.MyValueTaskSource`1.System.Threading.Tasks.Sources.IValueTaskSource<T>.GetResult(Int16 token)
   at Tmds.DBus.Protocol.DBusConnection.CallMethodAsync[T](MessageBuffer message, MessageValueReader`1 valueReader, Object state)
   at Tmds.DBus.Protocol.Connection.CallMethodAsync[T](MessageBuffer message, MessageValueReader`1 reader, Object readerState)
   at Avalonia.FreeDesktop.DBusIme.IBus.IBusX11TextInputMethod.Connect(String name)
   at Avalonia.FreeDesktop.DBusIme.DBusTextInputMethodBase.OnNameChange(Exception e, ValueTuple`3 args) (IBusX11TextInputMethod #32057793)`

i couldn't progress very much after these errors. Is there anyone out there that can guide me to the right direction?

**Comment by trivalik** [comment]
Here an outdated version of CefGlue:
https://github.com/OutSystems/WebView
